File: AutomationCore/Selenium/SeleniumWebElement.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using AutomationCore.Shared;
using AutomationCore.Shared.Exceptions;
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;

namespace AutomationCore.Selenium
{
    public class SeleniumWebElement : SeleniumWebObject, ITestableWebElement
    {
        public IWebElement _baseObject;

        public SeleniumWebElement(IWebElement baseObject)
        {
            this._baseObject = baseObject;
            this._searchContext = baseObject;
        }

        public void Clear()
        {
            _baseObject.Clear();
        }

        public void Type(string text)
        {
            _baseObject.SendKeys(text);
        }

        public void Type(ITestableWebElement element, string text)
        {
            element.Type(text);
        }

        public void Click()
        {
            _baseObject.Click();
        }

        public void Click(ITestableWebElement element)
        {
            element.Click();
        }

        public void Select(string item, bool isValue = false)
        {
            if (isValue)
            {
                new SelectElement(_baseObject).SelectByValue(item);
            }
            else
            {
                new SelectElement(_baseObject).SelectByText(item);
            }
        }

        public void Select(ITestableWebElement element, string item, bool isValue = false)
        {
            element.Select(item, isValue);
        }

        public bool WaitForAppear(string targetSubElement, int timeout)
        {
            Stopwatch watch = new Stopwatch();
            watch.Start();

            while ((watch.ElapsedMilliseconds / 1000) < timeout)
            {
                IList<ITestableWebElement> elementsFound = FindSubElements(targetSubElement);
                if (elementsFound.Any())
                {
                    bool allElementsAreVisible = true;

                    foreach (ITestableWebElement element in elementsFound)
                    {
                        allElementsAreVisible = allElementsAreVisible && element.IsDisplayed();
                    }

                    if (allElementsAreVisible)
                    {
                        watch.Stop();
                        return true;
                    }
                }
            }
            watch.Stop();
            return false;
        }

        public bool WaitForDisappear(string targetSubElement, int timeout)
        {
            Stopwatch watch = new Stopwatch();
            watch.Start();

            while ((watch.ElapsedMilliseconds / 1000) < timeout)
            {
                IList<ITestableWebElement> elementsFound = FindSubElements(targetSubElement);
                if (elementsFound.Any())
                {
                    bool allElementsAreHidden = true;

                    foreach (ITestableWebElement element in elementsFound)
                    {
                        allElementsAreHidden = allElementsAreHidden && !element.IsDisplayed();
                    }

                    if (allElementsAreHidden)
                    {
                        watch.Stop();
                        return true;
                    }
                }
            }
            watch.Stop();
            return false;
        }

        public bool IsSelected()
        {
            return _baseObject.Selected;
        }

        public void SetChecked(bool value)
        {
            if (_baseObject.Selected != value)
            {
                _baseObject.Click();
            }

            if (_baseObject.Selected != value)
            {
                _baseObject.SendKeys(Keys.Space);
            }

            if (_baseObject.Selected != value)
            {
               throw new ActionFailedException(string.Format("Failed to set checked value to {0} for element!", value));
            }
        }

        public void SetChecked(ITestableWebElement element, bool value)
        {
            element.SetChecked(value);
        }

        public void WaitForAttributeState(string targetSubElement, string attributeName, Func<string, bool> condition, int timeout)
        {
            Stopwatch watch = new Stopwatch();
            watch.Start();

            while ((watch.ElapsedMilliseconds / 1000) < timeout)
            {
                IList<ITestableWebElement> elementsFound = FindSubElements(targetSubElement);
                if (elementsFound.Any())
                {
                    bool allElementHaveState = true;

                    foreach (ITestableWebElement element in elementsFound)
                    {
                        allElementHaveState = allElementHaveState && condition(element.GetAttribute(attributeName));
                    }

                    if (allElementHaveState)
                    {
                        watch.Stop();
                        return;
                    }
                }
                else
                {
                    throw new ObjectNotFoundException(string.Format("Unable to find {0}, please check your object definitions.", targetSubElement));
                }
            }
            watch.Stop();
            throw new ActionTimeoutException(string.Format("Element {0} failed to get to the expected state in {1} seconds.", targetSubElement, timeout));
        }

        public bool IsDisplayed()
        {
            return _baseObject.Displayed;
        }

        public string GetAttribute(string attributeName)
        {
            return _baseObject.GetAttribute(attributeName);
        }

        public string GetTagName()
        {
            return _baseObject.TagName;
        }

        public string InnerHtml()
        {
            return GetAttribute("innerHTML");
        }

        public ITestableWebElement Parent(int levels = 1)
        {
            var xpath = "..";

            for (int i = 1; i < levels; i++)
            {
                xpath += "/..";
            }

            return new SeleniumWebElement(_baseObject.FindElement(By.XPath(xpath)));
        }
    }
}

File: js/src/Selenium/SeleniumWebElement.ts
/// <reference path="../../typings/selenium-webdriver.d.ts" />
/// <reference path="../core/interfaces.ts" />

class SeleniumWebElement implements Interfaces.ITestableWebElement  {

    constructor(_baseObject: webdriver.WebElement, DefaultActionTimeout?: number) {
        this._baseObject = _baseObject;
        this.DefaultActionTimeout = DefaultActionTimeout || 60;
        this.SubElements = {};
    }

    _baseObject: webdriver.WebElement;
    DefaultActionTimeout: number;
    SubElements: { [index: string]: { [index: string]: string; }; };

    Clear(): webdriver.promise.Promise<void> {
        this.EnsureElementFocus();
        return this._baseObject.clear();
    }

    Type(text: string): webdriver.promise.Promise<void>;
    Type(text: string, element: Interfaces.ITestableWebElement): webdriver.promise.Promise<void>;
    Type(text: string, element?: Interfaces.ITestableWebElement): webdriver.promise.Promise<void> {
        if (element == null) {
            return this._baseObject.sendKeys(text);
        } else {
            return element.Type(text);
        }
    }

    Click(): webdriver.promise.Promise<void>;
    Click(element: Interfaces.ITestableWebElement): webdriver.promise.Promise<void>;
    Click(element?: Interfaces.ITestableWebElement): webdriver.promise.Promise<void> {
        if (element == null) {
            return this._baseObject.click();
        } else {
            return element.Click();
        }
    }

    Select(item: string, isValue: boolean): webdriver.promise.Promise<void>;
    Select(item: string, isValue: boolean, element: Interfaces.ITestableWebElement): webdriver.promise.Promise<void>;
    Select(item: string, isValue: boolean, element?: Interfaces.ITestableWebElement): webdriver.promise.Promise<void> {
        

        var optionProperties: { [key: string]: string } = { "tag": "option" }
        if (isValue) {
            optionProperties["value"] = item;
        } else {
            optionProperties["text"] = item;
        }

        if (element == null) {
            return this.Click()
                    .then(() => this.FindSubElement(optionProperties))
                    .then((element) => element.Click());
        } else {
            return element.Click()
                    .then(() => element.FindSubElement(optionProperties))
                    .then((subElement) => subElement.Click());
        }
    }

    Exists(targetSubElement: string, timeout: number): webdriver.promise.Promise<boolean> {
        var condition = this.FindSubElements(targetSubElement).then((elementsFound) => {
            if (elementsFound.length > 0) {
                return true;
            }
            return false;
        });

        return this._baseObject.getDriver().wait(condition, timeout);
    }

    WaitForAppear(timeout: number): webdriver.promise.Promise<boolean>;
    WaitForAppear(timeout: number, targetSubElement: string): webdriver.promise.Promise<boolean>;
    WaitForAppear(timeout: number, targetSubElement?: string): webdriver.promise.Promise<boolean> {
        var condition: webdriver.promise.Promise<boolean>;

        if (targetSubElement == null) {
            condition = this.IsDisplayed();
        } else {
            condition = this.FindSubElements(targetSubElement).then((elementsFound) => {
                if (elementsFound.length > 0) {
                    //Create promise that finds all the elements on the page and returns if they are displayed or not
                    return webdriver.promise.all(elementsFound.map((element) => element.IsDisplayed()));
                }
                return new webdriver.promise.Promise((accept) => accept([false]));
            }).then((results) => { //Verify all elements are displayed
                return results.filter((item) => item === false).length === 0;
            });
        }

        return this._baseObject.getDriver().wait(condition, timeout);
    }

    WaitForDisappear(timeout: number): webdriver.promise.Promise<boolean>;
    WaitForDisappear(timeout: number, targetSubElement: string): webdriver.promise.Promise<boolean>;
    WaitForDisappear(timeout: number, targetSubElement?: string): webdriver.promise.Promise<boolean> {
        var condition: webdriver.promise.Promise<boolean>;

        if (targetSubElement == null) {
            condition = this.IsDisplayed().then((result) => !result);
        } else {
            var condition = this.FindSubElements(targetSubElement).then((elementsFound) => {
                if (elementsFound.length > 0) {
                    //Create promise that finds all the elements on the page and returns if they are displayed or not
                    return webdriver.promise.all(elementsFound.map((element) => element.IsDisplayed()));
                }
                return new webdriver.promise.Promise((accept) => accept([false]));
            }).then((results) => { //Verify all elements are not displayed
                return results.filter((item) => item !== false).length === 0;
            });
        }

        return this._baseObject.getDriver().wait(condition, timeout);
    }

    IsDisplayed(): webdriver.promise.Promise<boolean> {
        this.EnsureElementFocus();
        return this._baseObject.isDisplayed();
    }

    IsSelected(): webdriver.promise.Promise<boolean> {
        this.EnsureElementFocus();
        return this._baseObject.isSelected();
    }

    SetChecked(value: boolean): webdriver.promise.Promise<void>;
    SetChecked(value: boolean, element: Interfaces.ITestableWebElement): webdriver.promise.Promise<void>;
    SetChecked(value: boolean, element?: Interfaces.ITestableWebElement): webdriver.promise.Promise<void> {
        if (element != null) {
            return element.SetChecked(value);
        }

        this.EnsureElementFocus();
        
        return this.IsSelected()
                .then((selected) => { //Try clicking first
                    if (!selected) {
                        this.Click();
                    }
                    return this.IsSelected();
                }).then((selected) => { //Then try using space
                    if (!selected) {
                        this._baseObject.sendKeys(webdriver.Key.ENTER);
                    }
                    return this.IsSelected();
                }).then((selected) => { //If that fails throw
                    if (!selected) {
                        throw new Error("Failed to set element to checked!");
                    }
                });
    }

    WaitForAttributeState(attributeName: string, condition: (value: string) => boolean, timeout: number): webdriver.promise.Promise<boolean>;
    WaitForAttributeState(attributeName: string, condition: (value: string) => boolean, timeout: number, targetSubElement: string): webdriver.promise.Promise<boolean>;
    WaitForAttributeState(attributeName: string, condition: (value: string) => boolean, timeout: number, targetSubElement?: string): webdriver.promise.Promise<boolean> {
        var waitCondition: webdriver.promise.Promise<boolean>;

        if (targetSubElement == null) {
            waitCondition = this.GetAttribute(attributeName).then((value) => condition(value));
        } else {
            waitCondition = this.FindSubElements(targetSubElement).then((elementsFound) => {
                if (elementsFound.length > 0) {
                    return webdriver.promise.all(elementsFound.map((element) => element.GetAttribute(attributeName)));
                }
                return new webdriver.promise.Promise((accept) => accept([null]));
            }).then((results) => {
                return results.every((item) => item != null && condition(item));
            });
        }

        return this._baseObject.getDriver().wait(waitCondition, timeout);}

    GetAttribute(attributeName: string): webdriver.promise.Promise<string> {
        this.EnsureElementFocus();
        return this._baseObject.getAttribute(attributeName);
    }

    GetTagName(): webdriver.promise.Promise<string> {
        this.EnsureElementFocus();
        return this._baseObject.getTagName();
    }

    InnerHtml(): webdriver.promise.Promise<string> {
        this.EnsureElementFocus();
        return this._baseObject.getInnerHtml();
    }

    GetText(): webdriver.promise.Promise<string> {
        this.EnsureElementFocus();
        return this._baseObject.getText();
    }

    Parent(levels: number): webdriver.promise.Promise<SeleniumWebElement> {

        this.EnsureElementFocus();
        var xpath = "..";

        for (var i = 1; i < levels; i++)
        {
            xpath += "/..";
        }

        return this._baseObject.findElement(webdriver.By.xpath(xpath)).then((element) => new SeleniumWebElement(element));
    }

    //
    //Search Context Methods
    //

    EnsureElementFocus() {
        return;
    }

    GetElementProperties(targetElement: string): { [key: string]: string; } {
        if (this.SubElements[targetElement]) {
            return this.SubElements[targetElement];
        }
        throw new Error(`Element name ${targetElement} is not a registered element name.`);
    }

    FindSubElement(targetSubElement: string): webdriver.promise.Promise<SeleniumWebElement>;
    FindSubElement(targetSubElement: string, timeout: number): webdriver.promise.Promise<SeleniumWebElement>;
    FindSubElement(subElementProperties: { [index: string]: string; }): webdriver.promise.Promise<SeleniumWebElement>;
    FindSubElement(subElementProperties: { [index: string]: string; }, timeout: number): webdriver.promise.Promise<SeleniumWebElement>;
    FindSubElement(targetSubElement: any, timeout?: number): webdriver.promise.Promise<SeleniumWebElement> {
        this.EnsureElementFocus();

        //If argument is a string then the target was a pre-defined sub element
        if (typeof targetSubElement === "string") {
            return this.FindSubElementByProperties(this.GetElementProperties(targetSubElement), timeout);
        }

        var elementProperties: { [index: string]: string; };
        if (typeof targetSubElement === typeof elementProperties) {
            return this.FindSubElementByProperties(targetSubElement, timeout);
        }

        throw new Error(`Invalid target of type ${typeof targetSubElement} found for findSubElement, must be string or Dictionary: { [index: string]: string; }`);
    }

    FindSubElementByProperties(subElementProperties: { [index: string]: string; }, timeout?: number): webdriver.promise.Promise<SeleniumWebElement> {
        if (!!subElementProperties["parentelement"]) {
            var parentName = subElementProperties["parentelement"];
            var elementProperties: { [index: string]: string; } = {};

            for (var prop in subElementProperties) {
                if (subElementProperties.hasOwnProperty(prop)) {
                    elementProperties[prop] = subElementProperties[prop];
                }
            }
            delete elementProperties["parentelement"];
            
            //Compiler complains about this not being a webdriver.promise.Promise<SeleniumWebElement> when it actually is
            return <any>FindExtensions.FindSubElement(this._baseObject, this.GetElementProperties(parentName))
                .then((parent) => parent.FindSubElement(elementProperties, timeout));
        }

        var condition = FindExtensions.FindSubElement(this._baseObject, subElementProperties).then((elementFound) => elementFound, () => null);

        return this._baseObject.getDriver().wait(condition, timeout);
    }

    FindSubElements(targetSubElement: string): webdriver.promise.Promise<SeleniumWebElement[]>;
    FindSubElements(targetSubElement: string, timeout: number): webdriver.promise.Promise<SeleniumWebElement[]>;
    FindSubElements(subElementProperties: { [index: string]: string; }): webdriver.promise.Promise<SeleniumWebElement[]>;
    FindSubElements(subElementProperties: { [index: string]: string; }, timeout: number): webdriver.promise.Promise<SeleniumWebElement[]>;
    FindSubElements(targetSubElement: any, timeout?: number): webdriver.promise.Promise<SeleniumWebElement[]> {
        this.EnsureElementFocus();

        //If argument is a string then the target was a pre-defined sub element
        if (typeof targetSubElement === "string") {
            return this.FindSubElementsByProperties(this.GetElementProperties(targetSubElement), timeout);
        }

        var elementProperties: { [index: string]: string; };
        if (typeof targetSubElement === typeof elementProperties) {
            return this.FindSubElementsByProperties(targetSubElement, timeout);
        }

        throw new Error(`Invalid target of type ${typeof targetSubElement} found for findSubElements, must be string or Dictionary: { [index: string]: string; }`);
    }

    FindSubElementsByProperties(subElementProperties: { [index: string]: string; }, timeout?: number): webdriver.promise.Promise<SeleniumWebElement[]> {
        if (!!subElementProperties["parentelement"]) {
            var elementProperties: { [index: string]: string; } = {};

            for (var prop in subElementProperties) {
                if (subElementProperties.hasOwnProperty(prop)) {
                    elementProperties[prop] = subElementProperties[prop];
                }
            }
            delete elementProperties["parentelement"];
            return this.FindSubElementsByProperties(elementProperties, timeout);
        }

        var condition = FindExtensions.FindSubElements(this._baseObject, subElementProperties).then((elementsFound) => {
            if (elementsFound.length > 0) {
                return elementsFound;
            }
            return [];
        });

        return this._baseObject.getDriver().wait(condition, timeout);
    }

    RegisterSubElement(name: string, elementProperties: { [index: string]: string; }): void {
        this.SubElements[name] = {};

        //lowercase all element properties
        for (var prop in elementProperties) {
            if (elementProperties.hasOwnProperty(prop)) {
                this.SubElements[name][prop.toLowerCase()] = elementProperties[prop];
            }
        }
    }

}
